### ISSUE TITLE
Next semver/version tag?

### ISSUE THREAD

**Issue by ryan0x44** [prometheus/client_golang#344]
When using tools like [Dep](https://github.com/golang/dep) it defaults to the `v0.8.0` tag which was last tagged on Aug 17, 2016.

Is there a plan to tag a new release anytime soon? It would be great to not have to pin to a specific commit hash and use semver instead :)

**Comment by beorn7** [comment]
You can follow the progress towards v0.9 and v0.10 on the [milestones page](https://github.com/prometheus/client_golang/milestones).
There is no real formal release process. We don't offer any packages from out side. Semver prior to 1.0 is also not really strict, so nobody can be really sure what "use semver" means precisely prior to 1.0.

Overall, I'm not convinced about the benefits of tagging new incremental versions frequently. If there is no pressing reason, I would like to stick to the roadmap laid out on https://github.com/prometheus/client_golang/milestones . It was created before the Dep tool became a real thing. Before that, version tags did not really have a special meaning over any other commit ID.

Having said all that, I have set finishing 0.9 as my first priority for Prometheus work.

**Comment by zackbloom** [comment]
Understood, and we all appreciate your work on this (open source) project.

To answer your question on the benefit of tagging new incremental versions frequently, I think the answer is tools like dep look to these tags (if they're available). If you expect users to be using the bugfixes and features you added since v0.8.0, there is value in tagging those changes in releases. Otherwise we're either stuck using what's in v0.8.0, or jumping through hoops to manually override its default behavior.

**Comment by beorn7** [comment]
Hmmmm… So the original idea of Go was to always build against head, as Google does it internally, and as it is the basic idea in any monorepo world. Accordingly, we keep head of any Prometheus repo in working state always (or at least we aim to do so), so you could totally always build against head.

Now humanity realized that the world is not a monorepo, so vendoring kicked in. You freeze the state you build against because you might not be ready to go to the next commit yet because of some API changes. Semver is supposed to tell you when an incompatible change has happened. So far so good. However, prior to 1.0, semver has not really a meaning. I cannot increment the major version as it is pinned to 0. Both compatible feature changes as well as incompatible ones would only increment the minor version. In terms of bugs, client_golang is relatively stable. So I would essentially increment the minor version with (almost) every commit, as every commit gives us a new feature, compatible or not. If your vendoring tool defaults to the latest tag (or the latest that is not a major version change), there would essentially be no difference to building from head. But why are you vendoring then in the first place?

Bottom line is that I'm not convinced that frequent version tagging makes sense in a pre-1.0 semver context unless there is some kind of release process (because we package something somewhere or something).

**Comment by sdurrheimer** [comment]
Can [gopkg.in](http://labix.org/gopkg.in) help with that ?

**Comment by beorn7** [comment]
That would again just create semver semantics.

However, the whole reason for being pre-1.0 is that breaking changes happen all the time and you would be at 42.0.0 in no time if you actually applied semver with its intended 1.0+ semantics. It would also be a huge drain of maintenance time if you did this with a project still under heavy pre-1.0 development. As you can see looking at the descriptions of the milestones, I try to bundle "heavily" breaking changes into 0.10, while doing only "mildly" breaking changes for now. However, semver requires to increase the major version even for those. And since there are still many, we are pre-1.0.

I do think for a well-maintained pre-1.0 repo (i.e. one where HEAD always works), it's the sanest approach to pin to the current HEAD once you start using it in your own project. Then you occasionally update your pinning with caution should you need new features and/or bug fixes. Once the repo is at 1.0+, you can enjoy all the goodness of semver semantics. I would like a dependency management tool to have good support for that workflow.

However, the _fundamental_ problem here is that I am right now investing the time I can spent on this repo to discuss a problem that would not exist in the first place if I could spend my time on actually getting v0.9 ready. Thus, I took the pragmatic approach now and created an arbitrary [tag](https://github.com/prometheus/client_golang/releases/tag/v0.9.0-pre1) to make you dependency management tool happy.

**Comment by zackbloom** [comment]
Just to make the point, I don't think you would be at 42.0.0, because we're talking about using the patch version number for something meaningful. As you said, before 1.x.x, the third number can also represent minor or even breaking changes. This means you could release 0.8.1, 0.8.2, etc. and it would solve the issue.

I appreciate this is a pain, it always has been for me working with NPM projects, but it is how most package managers seem to refer to versions. In any case, thanks for your work, looking forward to 0.9.
